### ISSUE TITLE
Next v13 `app-build-manifest.json` -  Does not register a service worker that controls page and start_url

### ISSUE THREAD

**Issue by jpnws** [shadowwalker/next-pwa#424]
# Summary

I was running into "Does not register a service worker that controls page and start_url" after migrating to NextJS v13.

![image](https://user-images.githubusercontent.com/61026408/198233384-c3489e32-e57e-4078-8520-0ba8205bfeaa.png)

After some investigation, I noticed that I was running into a bad precaching runtime error due to `http://localhost:3000/_next/app-build-manifest.json`.

![image](https://user-images.githubusercontent.com/61026408/198233156-c5e80c64-7bc9-4330-b749-adce055b260b.png)

Does this have something to do with the fact that Next13 uses `app` directory?

*Versions:*

- `next-pwa`: 5.6.0
- `next`: 13

Steps to reproduce the behavior: Migrate to NextJS v13.

Expected Behaviors: Service worker should be succesfully registered.

**Comment by Oussamaosman02** [comment]
NextJs 13 still don´t support some other libraries like this one because they have to update it to the app folder.
If you want the pwa you have to do it with pages folder

**Comment by Rishab1207** [comment]
@jpnws are you adding your pages into the `app` directory?

**Comment by BluThaitanium** [comment]
Not a fix, but a behavior I found that gets PWA working briefly on apps not using `pages`:

1. Comment out appDir (breaks website)
2. `npm run build`, `npm run dev`
3. Open `localhost:3000`
- You should see 404 not found

4. Uncomment appDir
5. `npm run build`, `npm run dev`
6. Refresh webpage, and the PWA should work
- I did this on Chrome. PWA shows up working in `console` and `applications` tab in dev tools. Once I ran `Lighthouse`, things stop working again.

```
/** @type {import('next').NextConfig} */

const withPWA = require('next-pwa')({
  dest: 'public',
})

module.exports = withPWA({
  experimental: {
    appDir: true,             // <---- Comment and Uncomment this
  },
})
```

**Comment by jpnws** [comment]
> @jpnws are you adding your pages into the `app` directory?

Yes I'm using the new app directory on NextJS 13.

**Comment by DuCanhGH** [comment]
I ended up forking this repo: https://github.com/DuCanhGH/next-pwa (it works with App Router and has built-in TypeScript, JSDoc support - all examples are also written in TypeScript).

This happens because [this line](https://github.com/shadowwalker/next-pwa/blob/1e6af5fa9f6a807930f451adb5ab5078d5cda451/index.js?plain=1#L217) doesn't check for `app-build-manifest.json` alongside with `build-manifest.json` although it is supposed to as there hasn't been any activity on this repo since App Router was released.

One alternative is changing your `next-pwa` config:
```js
// next.config.js
const withPWAInit = require("next-pwa");

const isDev = process.env.NODE_ENV !== "production";

const withPWA = withPWAInit({
    // your other config...
    exclude: [
        // add buildExcludes here
        ({ asset, compilation }) => {
            if (
                asset.name.startsWith("server/") ||
                asset.name.match(/^((app-|^)build-manifest\.json|react-loadable-manifest\.json)$/)
            ) {
                return true;
            }
            if (isDev && !asset.name.startsWith("static/runtime/")) {
                return true;
            }
            return false;
        }
    ],
});

/** @type {import("next").NextConfig} */
const nextConfig = {
    // your other config...
    webpack(config) {
        const registerJs = path.join(path.dirname(require.resolve("next-pwa")), "register.js");
        const entry = config.entry;

        config.entry = () =>
            entry().then((entries) => {
                // Automatically registers the SW and enables certain `next-pwa` features in 
                // App Router (https://github.com/shadowwalker/next-pwa/pull/427)
                if (entries["main-app"] && !entries["main-app"].includes(registerJs)) {
                    if (Array.isArray(entries["main-app"])) {
                        entries["main-app"].unshift(registerJs);
                    } else if (typeof entries["main-app"] === "string") {
                        entries["main-app"] = [registerJs, entries["main-app"]];
                    }
                }
                return entries;
            });

        return config;
    },
}

module.exports = nextConfig;
```

Edit: updated the solution based on @Schular's comment.

**Comment by CombeeMike** [comment]
Does anyone know if there are any plans to also add this fix to this repo anytime soon?

**Comment by evolify** [comment]
Any update, or any alternative?

**Comment by AlainYRS** [comment]
Done! Create a head.tsx in app root directory as client component, then with a use effect hook register the service worker.

**Comment by AaronLayton** [comment]
@AlainYRS you got an example you could paste in?

**Comment by DuCanhGH** [comment]
@AaronLayton I don't think his solution would work. If you want to use next-pwa, you can try adding this to your next-pwa config:
```js
const withPWAInit = require("next-pwa");

const isDev = process.env.NODE_ENV === "production";

const withPWA = withPWAInit({
  // your other config
  exclude: [
    ({ asset, compilation }) => {
      if (
        asset.name.startsWith("server/") ||
        asset.name.match(/^(build-manifest\.json|react-loadable-manifest\.json)$/)
      ) {
        return true;
      }
      if (isDev && !asset.name.startsWith("static/runtime/")) {
        return true;
      }
      return false;
    }
  ],
});
```

**Comment by AaronLayton** [comment]
Thanks @DuCanhGH that works for me. I did try and exclude the `app-build-manifest` but think I was using `buildExcludes` instead of `exclude`. 

**Temporary fix**  
For anyone coming here for the same reason and before this PR gets merged - https://github.com/shadowwalker/next-pwa/pull/427 - the below works for me in production with the following caveat.

This excludes the nextjs build manifest which includes all the page URLs that get generated. In my testing, the below will pre cache static files but it won't cache any URL until you first visit the page **after** the service-worker has been installed on first visit. 

`next-pwa": "^5.6.0`

**next.config.js**
```tsx
const withPWAInit = require("next-pwa");

const isDev = process.env.NODE_ENV !== "production";

const withPWA = withPWAInit({
  dest: 'public',
  disable: process.env.NODE_ENV === 'development',
  
  exclude: [
    // add buildExcludes here
    ({ asset, compilation }) => {
      if (
        asset.name.startsWith("server/") ||
        asset.name.match(/^((app-|^)build-manifest\.json|react-loadable-manifest\.json)$/)
      ) {
        return true;
      }
      if (isDev && !asset.name.startsWith("static/runtime/")) {
        return true;
      }
      return false;
    }
  ],
});

/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    appDir: true,
  },
}

module.exports = withPWA(nextConfig);
```


**app/head.tsx**
```tsx
import HeadPwa from "./headPwa";

export default function Head() {
  return (
    <>
      <title>Your App Title</title>
      <meta charSet="utf-8" />
      <meta content="width=device-width, initial-scale=1" name="viewport" />
      <meta name="description" content="Some awesome app description" />
      <link rel="icon" href="/favicon.ico" />

      <HeadPwa />
    </>
  )
}
```

Custom PWA Header component that runs in the client

**app/headPwa.tsx**
```tsx
"use client"

import { useEffect } from "react";

export default function HeadPwa() {
  let sw: ServiceWorkerContainer | undefined;

  if (typeof window !== "undefined") {
    sw = window?.navigator?.serviceWorker;
  }

  useEffect(() => {
    if (sw) {
      sw.register("/sw.js", { scope: "/" }).then((registration) => {
        console.log("Service Worker registration successful with scope: ", registration.scope);
      }).catch((err) => {
        console.log("Service Worker registration failed: ", err);
      });
    }
  }, [sw]);

  return (
    <></>
  )
}
```

**Comment by DuCanhGH** [comment]
@AaronLayton I think `<HeadPwa />` should be put in your `app/layout.tsx` instead.

Also, `disable: process.env.NODE_ENV === 'development'` can be changed to `disable: isDev`. Just saying :)

**Comment by AaronLayton** [comment]
@DuCanhGH fully agree - yeah the HeadPwa only ended up being there due to iterations to get to this point. Adding it to the `layout.tsx` will be just as good. I have updated the snippet above.

By all means this is not a `fix` but it should get people by for the time being.

I did try to contact @shadowwalker on Twitter to see if they were getting notifications or if there is any way I can help to get the PR merged in, but no luck yet ☹️

**Comment by DuCanhGH** [comment]
@AaronLayton I'd also suggest changing your `app/Pwa.tsx` a bit:
```tsx
import type { Workbox } from "workbox-window";

declare global {
  interface Window {
    workbox: Workbox;
  }
}

export default function Pwa() {
  useEffect(() => {
    if (
      "serviceWorker" in navigator &&
      window.workbox !== undefined
    ) {
      const wb = window.workbox;
      wb.register();
    }
  }, []);
  return <></>;
}
```
I think this looks best :) I also think it's kinda hard to contact the author, let alone have him accept the PR as he hasn't been working on this project for months...

**Comment by Murkrage** [comment]
@DuCanhGH I see you've made lots of changes since you forked the repo. I'm inclined to go with your fork instead but unsure what kind of impact your changes have on the package.

**Comment by DuCanhGH** [comment]
@Murkrage so far I've done these changes to the package:
- Migrated to TS (examples are also in TS).
- Added support for appDir (app/_offline/page.tsx now also automatically enables offline fallback for document).
- Migrated from babel to swc (the original package used next/babel to build fallback/custom worker).
- Moved workbox's options from general plugin options to `PluginOptions.workboxOptions` (this change is why the package is at its 7.x right now, and if you have `// @ts-check` at the top of your `next.config.js` you should also be informed if you pass invalid options, for example having both GenerateSW-specfic and InjectManifest-specific options).
- Fixed grammar.
These are the ones I can pull off the top of my head :)

**Comment by hk86** [comment]
Had the same issue. @DuCanhGH You package solves the issue. Thank you!

**Comment by ncinme** [comment]
I was struggling to implement PWA in Next13 App directory. Thank you @DuCanhGH. Your solution worked like a charm. :)

**Comment by Schular** [comment]
For anyone that wants a working solution using only the `next-pwa` (keep in mind that this repo seems like is not maintained anymore), I created a sample [repository](https://github.com/Schular/next-with-pwa/) with a workaround: [next.config.js](https://github.com/Schular/next-with-pwa/blob/main/next.config.js)

The above repo is deployed on Vercel _(app & pages dirs concomitantly)_ with the following pages:
- **App**: [/home](https://next-with-pwa.vercel.app/)
- **Pages**: [/test-pages](https://next-with-pwa.vercel.app/test-pages)

This example is a simple Next.js starter project that should get ~100% on all Lighthouse categories.
![image](https://github.com/shadowwalker/next-pwa/assets/33543135/45ef227f-ae8d-44cb-be0a-f5bd98b78d70)
![image](https://github.com/shadowwalker/next-pwa/assets/33543135/ea7a74cc-1c9e-41ec-be52-12f1ce5f25d4)


Basically we exclude the manifest/server/dev files at build (https://github.com/shadowwalker/next-pwa/issues/424#issuecomment-1332258575) and we modify the webpack config with the custom `main-app` entry (https://github.com/shadowwalker/next-pwa/pull/427)

Thanks @DuCanhGH, [floatingdino](https://github.com/floatingdino)

**Comment by DuCanhGH** [comment]
@Schular nice, that sounds like a great solution! I've updated my own comment to match this repo as well.

**Comment by flaviobzs** [comment]
@Schular does this solution work with bashPath?

**Comment by Schular** [comment]
> 



> @Schular does this solution work with bashPath?

Yes, you should have no problems.
For example adding `basePath: "/test"` then you would have to update the path to all public assets to have the `/test` prefix (manifest.json, favicon.ico, images, etc.), other than that should work out-of-the-box.

**Comment by joniel** [comment]
> For anyone that wants a working solution using only the `next-pwa` (keep in mind that this repo seems like is not maintained anymore), I created a sample [repository](https://github.com/Schular/next-with-pwa/) with a workaround: [next.config.js](https://github.com/Schular/next-with-pwa/blob/main/next.config.js)
> 
> The above repo is deployed on Vercel _(app & pages dirs concomitantly)_ with the following pages:
> 
> * **App**: [/home](https://next-with-pwa.vercel.app/)
> * **Pages**: [/test-pages](https://next-with-pwa.vercel.app/test-pages)
> 
> This example is a simple Next.js starter project that should get ~100% on all Lighthouse categories. ![image](https://user-images.githubusercontent.com/33543135/238758206-45ef227f-ae8d-44cb-be0a-f5bd98b78d70.png) ![image](https://user-images.githubusercontent.com/33543135/238758075-ea7a74cc-1c9e-41ec-be52-12f1ce5f25d4.png)
> 
> Basically we add `buildExcludes: ["app-build-manifest.json"]` at `next-pwa` config, as the others are already excluded on the [`next-pwa` implementation](https://github.com/shadowwalker/next-pwa/blob/1e6af5fa9f6a807930f451adb5ab5078d5cda451/index.js?plain=1#L217) ([#424 (comment)](https://github.com/shadowwalker/next-pwa/issues/424#issuecomment-1332258575)) and we modify the webpack config with the custom `main-app` entry (#427)
> 
> Thanks @DuCanhGH, [floatingdino](https://github.com/floatingdino)

i have successed by this config code... 
but, after update package Next version(13.4.9), (by below command) 
'yarn upgrade-interactive --latest'

after update new version next.js, i have met next error code
(I'm novice in JS  T_T;)

`
> [PWA] Compile server
> [PWA] Compile server
> [PWA] Compile client (static)
> [PWA] Auto register service worker with: S:\dev_test\next-with-pwa\node_modules\next-pwa\register.js  
> [PWA] Service worker: S:\dev_test\next-with-pwa\public\sw.js
> [PWA]   url: /sw.js
> [PWA]   scope: /
> [PWA] Build in develop mode, cache and precache are mostly disabled. This means offline support is disabled, but you can continue developing other functions in service worker.
- event compiled client and server successfully in 1617 ms (20 modules)
- wait compiling...
- wait compiling /page (client and server)...
- error unhandledRejection: Error: Entry app-client-internals depends on main-app, but this entry was not found
    at Compilation.seal (S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:101818)
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:148814
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:99267
    at _next2 (eval at create (S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:13:28867), <anonymous>:16:1)
    at eval (eval at create (S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:13:28867), <anonymous>:42:1)
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:305473
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\neo-async\async.js:1:9430
    at Object.each (S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\neo-async\async.js:1:9608)

    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:305212
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\neo-async\async.js:1:9430
    at done (S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\neo-async\async.js:1:9742)       
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\neo-async\async.js:1:9372
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:302404
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:15876
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:15437
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:536929
    at S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:28:15906
    at eval (eval at create (S:\dev_test\next-with-pwa\node_modules\next\dist\compiled\webpack\bundle5.js:13:28867), <anonymous>:30:1)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
` 
package.json
`
{
  "name": "next-app-dir-with-pwa",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "next": "13.4.9",
    "next-pwa": "^5.6.0",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "webpack": "^5.88.1"
  },
  "devDependencies": {
    "eslint": "8.44.0",
    "eslint-config-next": "13.4.9"
  }
}
`

**Comment by joniel** [comment]
in my case, Next 13.4.9 `yarn build ` success, but `yarn dev` failed. (I'm novice in JS framework...)
I leave a error code below.
![image](https://github.com/shadowwalker/next-pwa/assets/933971/8bacb205-0766-44b3-85c1-48b3b4670688)

**Comment by NikitaPoly** [comment]
The above solutions worked for my next.js project using the `/src`  with `/app`. My next issue was configuring the start_url. My start url is not `/` so the service worker would not get registered. After installing the app and running it a 404 error would appear. Lighthouse test showed that no service worker was registered on the `start_url`. The solution I found that lets you add a `start_url` that is something other than `/`:

1. Use solutions from above to get the `start_url:/` to work.
2. Change manifest.json to point to correct `start_url`
I also put my manifest and icons into a separate worker directory in `/public` (optional)
If you did change the location of the manifest/icons then also make sure to update the routes of the icons in the manifest and the meta import for manifest.json 
3. Change next.config.js to include `sw:"nameOfWorker.js"` (Optional )
4. Register the service worker in a client component on you desried `start_url`
 ```
useEffect(() => {
    if ("serviceWorker" in navigator) {
      window.addEventListener("load", function () {
        navigator.serviceWorker.register("/nameofSW.js").then(
          function (registration) {
            console.log("Service Worker registration successful with scope: ", registration.scope);
          },
          function (err) {
            console.log("Service Worker registration failed: ", err);
          }
        );
      });
    }
  }, []);
  const [installPromptEvent, setInstallPromptEvent] = useState(null);
```
!!!The service worker is still at root of website so you don't need to prefix it like the `manifest.json`!!!
5. npm run build and start

**Comment by booster085** [comment]
> @AaronLayton I don't think his solution would work. If you want to use next-pwa, you can try adding this to your next-pwa config:
> 
> ```js
> const withPWAInit = require("next-pwa");
> 
> const isDev = process.env.NODE_ENV !== "production";
> 
> const withPWA = withPWAInit({
>   // your other config
>   exclude: [
>     // add buildExcludes here
>     ({ asset, compilation }) => {
>       if (
>         asset.name.startsWith("server/") ||
>         asset.name.match(/^((app-|^)build-manifest\.json|react-loadable-manifest\.json)$/)
>       ) {
>         return true;
>       }
>       if (isDev && !asset.name.startsWith("static/runtime/")) {
>         return true;
>       }
>       return false;
>     }
>   ],
> });
> 
> // your Next config
> ```
> 
> and you still have to register the SW yourself.

Thanks, this works for me too

**Comment by JWesorick** [comment]
After adding the workaround provided by @Schular I get the error `unhandledRejection: Error: Entry app/page depends on main-app, but this entry was not found` in the terminal.

using next@npm:14.1.0 [d369e]
next-pwa@npm:5.6.0 [d369e]
